apr: fix cross compile

squalus opened the pull request:
###### Description of changes
Added more answers to configure tests. Verified that the answers are the
same as the x86_64-linux and aarch64-linux native builds.

Added a patch to work around the /dev/zero mmap detection logic that
doesn't work in the cross case.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Mindavi commented:
These are default for linux x86_64?

squalus commented:
> These are default for linux x86_64?

The original PR didn't have enough answers to the configure tests. I went through and made sure the answers match the native builds for both x86_64-linux and aarch64-linux. Also had to add a patch to get it to match. (The patch explains in detail)

Mindavi reviewed (approved):
That's a lot of flags, but if they're needed there's no choice.

Mindavi commented:
I think we could apply the patch unconditionally. Shouldn't hurt for native.

squalus commented:
> That's a lot of flags, but if they're needed there's no choice.

Yeah. They are needed unfortunately to get the configure output to match the native build.

Mindavi commented:
@ofborg build pkgsCross.aarch64-multiplatform.apr

squalus commented:
Updated to unconditionally apply the patch.

squalus commented:
101-500 rebuilds. Under the threshold for staging?

Mindavi commented:
Yeah, <500 is master

b12f commented:
Result of `nixpkgs-review pr 188116` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>libsForQt512.kcachegrind</li>
    <li>libsForQt514.kcachegrind</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
  </ul>
</details>
<details>
  <summary>310 packages built:</summary>
  <ul>
    <li>adminer</li>
    <li>apacheHttpd (apacheHttpdPackages.apacheHttpd ,apacheHttpdPackages_2_4.apacheHttpd ,apacheHttpd_2_4)</li>
    <li>apacheHttpdPackages.mod_auth_mellon (apacheHttpdPackages_2_4.mod_auth_mellon)</li>
    <li>mod_ca (apacheHttpdPackages.mod_ca)</li>
    <li>mod_crl (apacheHttpdPackages.mod_crl)</li>
    <li>apacheHttpdPackages.mod_cspnonce (apacheHttpdPackages_2_4.mod_cspnonce)</li>
    <li>mod_csr (apacheHttpdPackages.mod_csr)</li>
    <li>mod_dnssd (apacheHttpdPackages.mod_dnssd)</li>
    <li>mod_fastcgi (apacheHttpdPackages.mod_fastcgi)</li>
    <li>apacheHttpdPackages.mod_itk (apacheHttpdPackages_2_4.mod_itk)</li>
    <li>apacheHttpdPackages.mod_mbtiles (apacheHttpdPackages_2_4.mod_mbtiles)</li>
    <li>mod_ocsp (apacheHttpdPackages.mod_ocsp)</li>
    <li>apacheHttpdPackages.mod_perl (apacheHttpdPackages_2_4.mod_perl)</li>
    <li>mod_pkcs12 (apacheHttpdPackages.mod_pkcs12)</li>
    <li>mod_python (apacheHttpdPackages.mod_python)</li>
    <li>mod_scep (apacheHttpdPackages.mod_scep)</li>
    <li>mod_spkac (apacheHttpdPackages.mod_spkac)</li>
    <li>mod_timestamp (apacheHttpdPackages.mod_timestamp)</li>
    <li>mod_wsgi (apacheHttpdPackages.mod_wsgi)</li>
    <li>apacheHttpdPackages.mod_wsgi3 (apacheHttpdPackages_2_4.mod_wsgi3)</li>
    <li>php (apacheHttpdPackages.php ,php81)</li>
    <li>apacheHttpdPackages.subversion (apacheHttpdPackages_2_4.subversion)</li>
    <li>apr</li>
    <li>aprutil</li>
    <li>arcanist</li>
    <li>bookstack</li>
    <li>cabal2nix</li>
    <li>dep2nix</li>
    <li>drush</li>
    <li>easyeffects</li>
    <li>engelsystem</li>
    <li>git-doc</li>
    <li>gitFull</li>
    <li>gitSVN</li>
    <li>gnome.gnome-control-center</li>
    <li>gnome.gnome-user-share</li>
    <li>haskellPackages.cabal2nix-unstable</li>
    <li>haxe (haxe_4_2)</li>
    <li>haxePackages.format</li>
    <li>haxePackages.heaps</li>
    <li>haxePackages.hlopenal</li>
    <li>haxePackages.hlsdl</li>
    <li>hxcpp (haxePackages.hxcpp)</li>
    <li>haxePackages.hxcs</li>
    <li>haxePackages.hxjava</li>
    <li>haxePackages.hxnodejs_4</li>
    <li>haxe_3_2</li>
    <li>haxe_3_4</li>
    <li>haxe_4_0</li>
    <li>haxe_4_1</li>
    <li>hydra_unstable</li>
    <li>icingaweb2</li>
    <li>libsForQt5.kcachegrind (libsForQt515.kcachegrind ,plasma5Packages.kcachegrind)</li>
    <li>log4cxx</li>
    <li>lsp-plugins</li>
    <li>mapcache</li>
    <li>mate.mate-user-share</li>
    <li>matomo</li>
    <li>matomo-beta</li>
    <li>modsecurity_standalone</li>
    <li>nagios</li>
    <li>neko</li>
    <li>nextcloud-news-updater</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-update-source</li>
    <li>nominatim</li>
    <li>pdepend</li>
    <li>perl534Packages.GoferTransporthttp</li>
    <li>perl534Packages.SVNSimple</li>
    <li>perl534Packages.libapreq2</li>
    <li>perl534Packages.mod_perl2</li>
    <li>perl536Packages.GoferTransporthttp</li>
    <li>perl536Packages.SVNSimple</li>
    <li>perl536Packages.libapreq2</li>
    <li>perl536Packages.mod_perl2</li>
    <li>phoronix-test-suite</li>
    <li>phosh</li>
    <li>php80</li>
    <li>php80Extensions.amqp</li>
    <li>php80Extensions.apcu</li>
    <li>php80Extensions.ast</li>
    <li>php80Extensions.bcmath</li>
    <li>php80Extensions.bz2</li>
    <li>php80Extensions.calendar</li>
    <li>php80Extensions.couchbase</li>
    <li>php80Extensions.ctype</li>
    <li>php80Extensions.curl</li>
    <li>php80Extensions.datadog_trace</li>
    <li>php80Extensions.dba</li>
    <li>php80Extensions.dom</li>
    <li>php80Extensions.ds</li>
    <li>php80Extensions.enchant</li>
    <li>php80Extensions.event</li>
    <li>php80Extensions.exif</li>
    <li>php80Extensions.ffi</li>
    <li>php80Extensions.fileinfo</li>
    <li>php80Extensions.filter</li>
    <li>php80Extensions.ftp</li>
    <li>php80Extensions.gd</li>
    <li>php80Extensions.gettext</li>
    <li>php80Extensions.gmp</li>
    <li>php80Extensions.gnupg</li>
    <li>php80Extensions.grpc</li>
    <li>php80Extensions.iconv</li>
    <li>php80Extensions.igbinary</li>
    <li>php80Extensions.imagick</li>
    <li>php80Extensions.imap</li>
    <li>php80Extensions.intl</li>
    <li>php80Extensions.ldap</li>
    <li>php80Extensions.mailparse</li>
    <li>php80Extensions.maxminddb</li>
    <li>php80Extensions.mbstring</li>
    <li>php80Extensions.memcached</li>
    <li>php80Extensions.mongodb</li>
    <li>php80Extensions.mysqli</li>
    <li>php80Extensions.mysqlnd</li>
    <li>php80Extensions.oci8</li>
    <li>php80Extensions.opcache</li>
    <li>php80Extensions.openssl</li>
    <li>php80Extensions.openswoole</li>
    <li>php80Extensions.pcntl</li>
    <li>php80Extensions.pcov</li>
    <li>php80Extensions.pdlib</li>
    <li>php80Extensions.pdo</li>
    <li>php80Extensions.pdo_dblib</li>
    <li>php80Extensions.pdo_mysql</li>
    <li>php80Extensions.pdo_oci</li>
    <li>php80Extensions.pdo_odbc</li>
    <li>php80Extensions.pdo_pgsql</li>
    <li>php80Extensions.pdo_sqlite</li>
    <li>php80Extensions.pdo_sqlsrv</li>
    <li>php80Extensions.pgsql</li>
    <li>php80Extensions.pinba</li>
    <li>php80Extensions.posix</li>
    <li>php80Extensions.protobuf</li>
    <li>php80Extensions.pspell</li>
    <li>php80Extensions.rdkafka</li>
    <li>php80Extensions.readline</li>
    <li>php80Extensions.redis</li>
    <li>php80Extensions.session</li>
    <li>php80Extensions.shmop</li>
    <li>php80Extensions.simplexml</li>
    <li>php80Extensions.smbclient</li>
    <li>php80Extensions.snmp</li>
    <li>php80Extensions.snuffleupagus</li>
    <li>php80Extensions.soap</li>
    <li>php80Extensions.sockets</li>
    <li>php80Extensions.sodium</li>
    <li>php80Extensions.sqlite3</li>
    <li>php80Extensions.sqlsrv</li>
    <li>php80Extensions.swoole</li>
    <li>php80Extensions.sysvmsg</li>
    <li>php80Extensions.sysvsem</li>
    <li>php80Extensions.sysvshm</li>
    <li>php80Extensions.tidy</li>
    <li>php80Extensions.tokenizer</li>
    <li>php80Extensions.xdebug</li>
    <li>php80Extensions.xml</li>
    <li>php80Extensions.xmlreader</li>
    <li>php80Extensions.xmlwriter</li>
    <li>php80Extensions.xsl</li>
    <li>php80Extensions.yaml</li>
    <li>php80Extensions.zend_test</li>
    <li>php80Extensions.zip</li>
    <li>php80Extensions.zlib</li>
    <li>php80Packages.box</li>
    <li>php80Packages.composer</li>
    <li>php80Packages.deployer</li>
    <li>php80Packages.grumphp</li>
    <li>php80Packages.phing</li>
    <li>php80Packages.phive</li>
    <li>php80Packages.php-cs-fixer</li>
    <li>php80Packages.php-parallel-lint</li>
    <li>php80Packages.phpcbf</li>
    <li>php80Packages.phpcs</li>
    <li>php80Packages.phpmd</li>
    <li>php80Packages.phpstan</li>
    <li>php80Packages.psalm</li>
    <li>php80Packages.psysh</li>
    <li>php81Extensions.amqp</li>
    <li>php81Extensions.apcu</li>
    <li>php81Extensions.ast</li>
    <li>php81Extensions.bcmath</li>
    <li>php81Extensions.bz2</li>
    <li>php81Extensions.calendar</li>
    <li>php81Extensions.couchbase</li>
    <li>php81Extensions.ctype</li>
    <li>php81Extensions.curl</li>
    <li>php81Extensions.datadog_trace</li>
    <li>php81Extensions.dba</li>
    <li>php81Extensions.dom</li>
    <li>php81Extensions.ds</li>
    <li>php81Extensions.enchant</li>
    <li>php81Extensions.event</li>
    <li>php81Extensions.exif</li>
    <li>php81Extensions.ffi</li>
    <li>php81Extensions.fileinfo</li>
    <li>php81Extensions.filter</li>
    <li>php81Extensions.ftp</li>
    <li>php81Extensions.gd</li>
    <li>php81Extensions.gettext</li>
    <li>php81Extensions.gmp</li>
    <li>php81Extensions.gnupg</li>
    <li>php81Extensions.grpc</li>
    <li>php81Extensions.iconv</li>
    <li>php81Extensions.igbinary</li>
    <li>php81Extensions.imagick</li>
    <li>php81Extensions.imap</li>
    <li>php81Extensions.intl</li>
    <li>php81Extensions.ldap</li>
    <li>php81Extensions.mailparse</li>
    <li>php81Extensions.maxminddb</li>
    <li>php81Extensions.mbstring</li>
    <li>php81Extensions.memcached</li>
    <li>php81Extensions.mongodb</li>
    <li>php81Extensions.mysqli</li>
    <li>php81Extensions.mysqlnd</li>
    <li>php81Extensions.oci8</li>
    <li>php81Extensions.opcache</li>
    <li>php81Extensions.openssl</li>
    <li>php81Extensions.openswoole</li>
    <li>php81Extensions.pcntl</li>
    <li>php81Extensions.pcov</li>
    <li>php81Extensions.pdlib</li>
    <li>php81Extensions.pdo</li>
    <li>php81Extensions.pdo_dblib</li>
    <li>php81Extensions.pdo_mysql</li>
    <li>php81Extensions.pdo_oci</li>
    <li>php81Extensions.pdo_odbc</li>
    <li>php81Extensions.pdo_pgsql</li>
    <li>php81Extensions.pdo_sqlite</li>
    <li>php81Extensions.pdo_sqlsrv</li>
    <li>php81Extensions.pgsql</li>
    <li>php81Extensions.pinba</li>
    <li>php81Extensions.posix</li>
    <li>php81Extensions.protobuf</li>
    <li>php81Extensions.pspell</li>
    <li>php81Extensions.rdkafka</li>
    <li>php81Extensions.readline</li>
    <li>php81Extensions.redis</li>
    <li>php81Extensions.session</li>
    <li>php81Extensions.shmop</li>
    <li>php81Extensions.simplexml</li>
    <li>php81Extensions.smbclient</li>
    <li>php81Extensions.snmp</li>
    <li>php81Extensions.snuffleupagus</li>
    <li>php81Extensions.soap</li>
    <li>php81Extensions.sockets</li>
    <li>php81Extensions.sodium</li>
    <li>php81Extensions.sqlite3</li>
    <li>php81Extensions.sqlsrv</li>
    <li>php81Extensions.swoole</li>
    <li>php81Extensions.sysvmsg</li>
    <li>php81Extensions.sysvsem</li>
    <li>php81Extensions.sysvshm</li>
    <li>php81Extensions.tidy</li>
    <li>php81Extensions.tokenizer</li>
    <li>php81Extensions.xdebug</li>
    <li>php81Extensions.xml</li>
    <li>php81Extensions.xmlreader</li>
    <li>php81Extensions.xmlwriter</li>
    <li>php81Extensions.xsl</li>
    <li>php81Extensions.yaml</li>
    <li>php81Extensions.zend_test</li>
    <li>php81Extensions.zip</li>
    <li>php81Extensions.zlib</li>
    <li>php81Packages.box</li>
    <li>php81Packages.composer</li>
    <li>php81Packages.deployer</li>
    <li>php81Packages.grumphp</li>
    <li>php81Packages.phing</li>
    <li>php81Packages.phive</li>
    <li>php81Packages.php-cs-fixer</li>
    <li>php81Packages.php-parallel-lint</li>
    <li>php81Packages.phpcbf</li>
    <li>php81Packages.phpcs</li>
    <li>php81Packages.phpmd</li>
    <li>php81Packages.phpstan</li>
    <li>php81Packages.psalm</li>
    <li>php81Packages.psysh</li>
    <li>platformsh</li>
    <li>pulseeffects-legacy</li>
    <li>python310Packages.pysvn</li>
    <li>ydiff (python310Packages.ydiff)</li>
    <li>python39Packages.pysvn</li>
    <li>python39Packages.ydiff</li>
    <li>qcachegrind</li>
    <li>rabbitvcs</li>
    <li>rapidsvn</li>
    <li>redwax-tool</li>
    <li>rss-bridge-cli</li>
    <li>serf</li>
    <li>simgear</li>
    <li>snipe-it</li>
    <li>subversion</li>
    <li>subversionClient</li>
    <li>svn-all-fast-export</li>
    <li>svn2git</li>
    <li>svnfs</li>
    <li>tests.php.overrideAttrs-preserves-enabled-extensions</li>
    <li>tests.php.withExtensions-enables-previously-disabled-extensions</li>
    <li>tomcat-native</li>
    <li>tomcat_connectors</li>
    <li>vcstool</li>
    <li>wp-cli</li>
    <li>wp4nix</li>
    <li>xsecurelock</li>
    <li>yle-dl</li>
    <li>zkfuse</li>
  </ul>
</details>

risicle commented on pkgs/development/libraries/apr/default.nix:
```diff
@@ -27,6 +27,10 @@ stdenv.mkDerivation rec {
       sha256 = "0hhm5v5wx985c28dq6d9ngnyqihpsphq4mw7rwylk39k2p90ppcm";
     })
+    (fetchpatch {
+      url = "https://830833.bugs.gentoo.org/attachment.cgi?id=761676";
```
```suggestion
      name = "cross-assume-dev-zero-mmappable.patch";
      url = "https://830833.bugs.gentoo.org/attachment.cgi?id=761676";
```

squalus replied:
Added the suggested name.

SuperSandro2000 commented on pkgs/development/libraries/apr/default.nix:
```diff
@@ -27,6 +27,10 @@ stdenv.mkDerivation rec {
+      url = "https://830833.bugs.gentoo.org/attachment.cgi?id=761676";
+      sha256 = "sha256-rsouJp1o7p0d+AJ5KvyhUU79vAJOcVHEuwSEKaCEGa8=";
+    })
   ] ++ lib.optionals stdenv.isDarwin [ ./is-this-a-compiler-bug.patch ];
```
Do you know by chance if this patch is still required and could be made unconditional?

squalus replied:
Yes, it is required, for the reasons discussed here: https://bugs.gentoo.org/show_bug.cgi?id=830833. The cross build will fail without it.

It is currently unconditional, as in, it's always applied.

squalus replied:
Oh sorry, I think you're talking about `is-this-a-compiler-bug.patch`. I can look into that.

squalus replied:
Added in a600443bb30e4895bf9d05ba309578be6f8ab951

squalus replied:
It's 7 years old, and no explanation about why it was required. I tested it out on darwin and the build works fine without the patch.

SuperSandro2000 replied:
If the build works fine without it I would just remove it.

SuperSandro2000 commented on pkgs/development/libraries/apr/default.nix:
```diff
@@ -44,8 +48,24 @@ stdenv.mkDerivation rec {
     "ac_cv_file__dev_zero=yes"
-    "ac_cv_func_setpgrp_void=0"
-    "apr_cv_tcp_nodelay_with_cork=1"
+    "ac_cv_func_setpgrp_void=yes"
```
We really should add a comment why we set all those options.

squalus replied:
Added an extra comment about the configure test answers.

squalus committed:
apr: fix cross compile

Added more answers to configure tests. Verified that the answers are the
same as the x86_64-linux and aarch64-linux native builds.

Added a patch to work around the /dev/zero mmap detection logic that
doesn't work in the cross case.

SuperSandro2000 commented on pkgs/development/libraries/apr/default.nix:
```diff
@@ -27,6 +27,13 @@ stdenv.mkDerivation rec {
+      name = "cross-assume-dev-zero-mmappable.patch";
+      sha256 = "sha256-rsouJp1o7p0d+AJ5KvyhUU79vAJOcVHEuwSEKaCEGa8=";
+    })
   ] ++ lib.optionals stdenv.isDarwin [ ./is-this-a-compiler-bug.patch ];
```
```suggestion
  ];
```

SuperSandro2000 committed:
apr: remove no longer required darwin patch
